### PR TITLE
Add Redemption Fee

### PR DESCRIPTION
**Pull request by sentilesdal**
Adds a redemption fee parameter to the repo.  This ended up touching a lot of files, the commits tell a story, but I think for this one it actually is easier to just look at the overall diff.  I was able to fix broken tests.  I've set the redemption fee to 0 for the calc_in/out tests for now and I'll address those specifically in a follow up PR since I'll have to crunch the numbers again.

Currently looking into why the market tests are broken.

### CODE REVIEW

File: src/elfpy/markets.py [hunk 1]
```diff
@@ -45,13 +45,15 @@ def __init__(
...
+        trade_fee_percent: float = 0,
+        redemption_fee_percent: float = 0,
     ):
         # market state variables
         self.time: float = 0  # t: timefrac unit is time normalized to 1 year, i.e. 0.5 = 1/2 year
         self.pricing_model = pricing_model
         self.market_state: MarketState = market_state
-        self.fee_percent: float = fee_percent  # g
+        self.trade_fee_percent: float = trade_fee_percent  # g
+        self.redemption_fee_percent: float = redemption_fee_percent  # g
```
Comment (dpaiton) [inline src/elfpy/markets.py:56]: this is not `g` -- pls fix or remove the comment

File: src/elfpy/pricing_models/hyperdrive.py [hunk 2]
```diff
@@ -35,7 +35,8 @@ def calc_in_given_out(
         self,
         out: Quantity,
         market_state: MarketState,
-        fee_percent: float,
+        trade_fee_percent: float,
+        redemption_fee_percent: float,
```
Comment (dpaiton) [inline src/elfpy/pricing_models/hyperdrive.py:39]: Can you update the docstrings? Probably here, in base, and in yieldspace.

File: src/elfpy/pricing_models/hyperdrive.py [hunk 3]
```diff
@@ -155,10 +161,10 @@ def calc_in_given_out(
             user_result=user_result,
             market_result=market_result,
             breakdown=TradeBreakdown(
-                without_fee_or_slippage=float(flat + Decimal(curve.breakdown.without_fee_or_slippage)),
-                without_fee=float(flat + Decimal(curve.breakdown.without_fee)),
-                fee=curve.breakdown.fee,
-                with_fee=float(flat + Decimal(curve.breakdown.with_fee)),
+                without_fee_or_slippage=float(flat_without_fee + Decimal(curve.breakdown.without_fee_or_slippage)),
+                without_fee=float(flat_with_fee + Decimal(curve.breakdown.without_fee)),
+                fee=curve.breakdown.fee + float(redemption_fee),
```
Comment (dpaiton) [inline src/elfpy/pricing_models/hyperdrive.py:166]: This has the opposite pattern as the other parts. Everywhere else you say `float(flat + Decimal(curve))` but here you say `curve + float(flat)`. Is that a typo, if not why is it different here?

File: src/elfpy/markets.py (hunk 1 above)
Comment (sentilesdal) [reply src/elfpy/markets.py:56]: its not, neither is trade fee, which is phi.  updated the comments

File: src/elfpy/pricing_models/hyperdrive.py (hunk 3 above)
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:166]: Not sure I understand, both calc_in_given_out and calc_out_given_in follow the same pattern for returning the trade result

File: src/elfpy/pricing_models/yieldspace.py [hunk 4]
```diff
@@ -323,7 +325,7 @@ def calc_in_given_out(
             and the unit of the tokens).
         market_state : MarketState
             The state of the AMM's reserves and share prices.
-        fee_percent : float
+        trade_fee_percent : float
             The percentage of the difference between the amount paid without
             slippage and the amount received that will be added to the input
             as a fee.
```
Comment (dpaiton) [inline src/elfpy/pricing_models/yieldspace.py:331]: Can you add `redemption_fee_percent` to the docstring pls?

File: src/elfpy/pricing_models/yieldspace.py [hunk 5]
```diff
@@ -545,7 +549,7 @@ def calc_out_given_in(
...
```
Comment (dpaiton) [inline src/elfpy/pricing_models/yieldspace.py:555]: Again add `redemption_fee_percent` to docstring pls

File: src/elfpy/utils/config.py [hunk 6]
```diff
@@ -182,7 +182,8 @@ def get_random_variables(config: Config):
         target_pool_apr=config.simulator.rng.uniform(
             low=config.amm.min_pool_apr, high=config.amm.max_pool_apr
         ),  # starting fixed apr as a decimal
-        fee_percent=config.simulator.rng.uniform(low=config.amm.min_fee, high=config.amm.max_fee),
+        trade_fee_percent=config.simulator.rng.uniform(low=config.amm.min_fee, high=config.amm.max_fee),
+        redemption_fee_percent=config.simulator.rng.uniform(low=config.amm.min_fee, high=config.amm.max_fee),
```
Comment (dpaiton) [inline src/elfpy/utils/config.py:186]: Are you sure that these two fees should have a single config parameter defining their range?

File: src/elfpy/pricing_models/hyperdrive.py [hunk 7]
```diff
@@ -278,9 +294,9 @@ def calc_out_given_in(
             user_result=user_result,
             market_result=market_result,
             breakdown=TradeBreakdown(
-                without_fee_or_slippage=float(flat + Decimal(curve.breakdown.without_fee_or_slippage)),
-                without_fee=float(flat + Decimal(curve.breakdown.without_fee)),
-                fee=curve.breakdown.fee,
-                with_fee=float(flat + Decimal(curve.breakdown.with_fee)),
+                without_fee_or_slippage=float(flat_without_fee + Decimal(curve.breakdown.without_fee_or_slippage)),
+                without_fee=float(flat_without_fee + Decimal(curve.breakdown.without_fee)),
+                fee=curve.breakdown.fee + float(redemption_fee),
```
Comment (dpaiton) [inline src/elfpy/pricing_models/hyperdrive.py:299]: also should be `float(redemption_fee) + curve.breakdown.fee)`

File: src/elfpy/utils/config.py (hunk 6 above)
Comment (sentilesdal) [reply src/elfpy/utils/config.py:186]: No I'm not sure.  We can change this if needed.

File: src/elfpy/pricing_models/hyperdrive.py (hunk 3 above)
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:168]: Ah I see what you mean now from IRL convo, changing.

File: src/elfpy/pricing_models/yieldspace.py (hunk 4 above)
Comment (sentilesdal) [reply src/elfpy/pricing_models/yieldspace.py:331]: yup!

File: src/elfpy/pricing_models/yieldspace.py (hunk 5 above)
Comment (sentilesdal) [reply src/elfpy/pricing_models/yieldspace.py:555]: yup!

File: src/elfpy/pricing_models/hyperdrive.py (hunk 7 above)
Comment (sentilesdal) [reply src/elfpy/pricing_models/hyperdrive.py:299]: roger

**Review by dpaiton — APPROVED** [review]
thanks!

**Review by jrhea — APPROVED** [review]
lgtm

**Pull request merged**